explorer/interpreter/type_checker.h

```diff
@@ -37,6 +37,22 @@ using GlobalMembersMap =
 
 class TypeChecker {
  public:
+  enum class PatternRefutability {
```

Comment:
I keep tying myself in knots trying to figure out things like whether "refutable" and "irrefutable" are mutually exclusive, and whether there's a superset/subset relationship among the enumerators (this could matter for figuring out whether there might be missing cases in `TypeCheckPattern`), and if so in which direction (which depends if we're thinking of them as sets of _requirements_ or sets of _patterns that have some property_).

To clarify all that, here's my suggestion:
- Document that values of this type represent a set of requirements that `TypeCheckPattern` must impose on the pattern.
- Rename the type `PatternRequirements`.
- Rename `Refutable` to `None` (because it imposes no requirements).
- Rename `BindingType` to `NoBindings`, and document it as a requirement that the pattern not contain binding subpatterns (with a TODO noting that the cases that need this should probably be expressions rather than patterns).

Reply:
Done, except for `NoBindings`.

> Rename `BindingType` to `NoBindings`, and document it as a requirement that the pattern not contain binding subpatterns (with a TODO noting that the cases that need this should probably be expressions rather than patterns).

In order to discourage use for other purposes, I prefer the purpose-specific name. I think `NoBindings` sounds more generic. I believe removing the enum value would be the right action, as already reflected by the TODO around #2788.

There is only one use and we're already trying to get rid of it, so I think it's better to be explicit about the intent that this is about #2788, rather than more broadly saying "cases that need this". If more uses are added, I think that's the point where more generic phrasing should be put in to use -- or the additional uses should just be refused, making the refactoring blocking.